### PR TITLE
Add deadlines to storage RPC calls

### LINKED ISSUE #4887: In distributed deployment, Minio responses slowly after one node with data is down
Following "minio server --help", set up a simple 4-node distributed minio, each node consists of 1 disk. Push some data to it, then shutdown a node, minio responses very slowly either from browser or minio client.

## Expected Behavior
minio should work as normally after 1 node shutdown since we still have (4/2 + 1 = 3) quorum.

## Current Behavior
Before shutting down the node:
```
docker@docker-ee-54 ~]$ time mc ls test54/dtr/docker
[2017-09-04 16:46:55 CST]     0B registry/
real	0m0.017s
user	0m0.004s
sys	0m0.005s
```
After shutting down the node, can not get response after 4 minutes
```
[docker@docker-ee-54 ~]$ time mc ls test54/dtr/docker
^C
real	4m7.454s
user	0m0.006s
sys	0m0.003s
```

## Possible Solution
Hope minio can identify and remove the dead node quickly.

## Steps to Reproduce (for bugs)
1. set up minio on 4 nodes
```
export MINIO_ACCESS_KEY=minio
export MINIO_SECRET_KEY=miniostorage
nohup /usr/local/bin/minio server \
    http://192.168.105.54/data/ \
    http://192.168.105.55/data/ \
    http://192.168.105.56/data/ \
    http://192.168.105.57/data/ > /tmp/minio.log 2>&1 &
```
2. Push some data (I'm using minio as Docker DTR storage backend, the bucket name is "dtr")
```
[docker@docker-ee-51 ~]$ docker push 192.168.105.51:8443/admin/alpine
The push refers to a repository [192.168.105.51:8443/admin/alpine]
5bef08742407: Pushed 
latest: digest: sha256:0930dd4cc97ed5771ebe9be9caf3e8dc5341e0b5e32e8fb143394d7dfdfa100e size: 528
```

3. Shutdown one node
```
[docker@docker-ee-57 foo]$ sudo shutdown -h now
```
4. Observe behavior of minio
Minio starts to response very slowly when accessing objects under "dtr" bucket.
In addition, pushing new Docker images such as "nginx" received errors.
```
[docker@docker-ee-51 ~]$ docker push 192.168.105.51:8443/admin/nginx:stable-alpine
The push refers to a repository [192.168.105.51:8443/admin/nginx]
1690bc77acd5: Retrying in 1 second 
cbeb94c1f91a: Retrying in 1 second 
4bbeb364e643: Retrying in 1 second 
040fd7841192: Retrying in 1 second 
received unexpected HTTP status: 504 Gateway Time-out
```
From dtr-registry-xxx container logs, I can see 
```
10.1.0.8 - - [04/Sep/2017:06:00:52 +0000] "HEAD /v2/admin/alpine/blobs/sha256:7328f6f8b41890597575cbaadc884e7386ae0acc53b747401ebce5cf0d624560 HTTP/1.1" 404 157 "" "docker/17.06.1-ee-2 go/go1.8.3 git-commit/8e43158 kernel/3.10.0-514.26.1.el7.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.1-ee-2 \\(linux\\))"t be enabled. Not submitting scan for repo: admin/nginx","time":"2017-09-04T08:23:46.19330931Z"}
```
While there were some data relating to "nginx" image uploaded to minio data directory (after shutting down 1 minio node).
```
[docker@docker-ee-55 ~]$ ls -ltr /data/dtr/docker/registry/v2/repositories/admin/nginx/_uploads/
total 0
drwxrwxr-x. 4 docker docker 41 Sep  4 14:59 e46b6069-419a-46d3-bd59-32a7e689a36a
drwxrwxr-x. 4 docker docker 41 Sep  4 15:00 bd08e3c4-ee65-423c-b1b8-63f0a86a9f2c
drwxrwxr-x. 4 docker docker 41 Sep  4 15:01 b56c3983-fe77-4a9b-8f06-016c48306a44
drwxrwxr-x. 4 docker docker 41 Sep  4 15:02 d2f70a30-5b66-4190-9343-6a36fd3a8cb2
drwxrwxr-x. 4 docker docker 41 Sep  4 15:07 cd98738d-b255-4cf6-b292-c91a901e36fe
....
```

## Context
I'm evaluating minio as a HA storage for Docker DTR.

## Your Environment

* Version used:
```
[docker@docker-ee-55 ~]$ minio version
Version: 2017-08-05T00:00:53Z
Release-Tag: RELEASE.2017-08-05T00-00-53Z
Commit-ID: aeafe668d8b6d25caac671d59e2b0f0473ce35d0
```
* Environment name and version 
```Docker Version:      17.06.1-ee-2, DTR 2.3.2, UCP-2.2.2```
* Server type and version:

* Operating System and version (`uname -a`):
```
CentOS Linux release 7.3.1611 (Core)
Linux docker-ee-55 3.10.0-514.26.1.el7.x86_64 #1 SMP Thu Jun 29 16:05:25 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
* Link to your project:

### LINKED ISSUE #5092: Minio unresponsive after shutting down node in a 4 node docker swarm
Created a 4 node cluster like https://github.com/minio/minio/tree/master/docs/orchestration/docker-swarm with a minio container pinned on each node. A traefik loadbalancer docker service in put in front of them.

When shutdown a node the minio system becomes unresponsive. When I just kill a minio container (not the whole host) everything works as expected. I've tested this through the load balancer and circumventing the load balancer by directly connecting with one of the alive nodes. I used the version from the guide and also the edge version, both did not work

It looks similar to the issue https://github.com/minio/minio/issues/4887 (and maybe also related to https://github.com/minio/minio/issues/5089), but the circuit-breaker workaround didnt work for me.

Slack messages: https://minio.slack.com/archives/C3NDUB8UA/p1508502645000199

### LINKED ISSUE #5290: If one out of three AWS instances goes down, then the Minio cluster becomes unresponsive
I have a setup with three EC2 instances, each with 1 docker container with 2 disks.
The cluster starts successfully and I can perform operations on it (using `mc`).
If I `docker kill` a node, then the cluster continue to work, because just 2 out of 6 disks are down.
If I terminate the EC2 instance (from the AWS console UI - stop or terminate), then the cluster becomes unresponsive. I cannot perform any `mc` operations on it (e.q. `mc ls <<alias>>` or `mc admin info <<alias>>). The command just return that the cluster if offline after several minutes. 

Please find bellow how each docker instance is configured:
`
- name: Storage node | Start storage node containers
  docker_container:
    name: "{{ item.1.docker_node_name }}"
    image: "minio/minio:RELEASE.2017-11-22T19-55-46Z"
    detach: true
    ports:
      - "9000:9000"
    restart_policy: always
    network_mode: "host"
    ulimits: nofile:100000:100000
    env:
      MINIO_ACCESS_KEY: "minio"
      MINIO_SECRET_KEY: "minio123"
    command: server http://minio1/export1 http://minio1/export2 http://minio2/export1 http://minio2/export2 http://minio3/export1 http://minio3/export2
  with_indexed_items: "{{ containers_for_host | json_query(filter_query) }}"
  when: item.1.type == 'storage'
  vars:
    filter_query: "[?ip=='{{ inventory_hostname }}'].containers"
    docker_node_name: "{{ item.1.docker_node_name }}"
  become: yes
  become_method: sudo
`

If I start the EC2 instance that was stopped/terminated, then the cluster detects this and becomes back online.

## Expected Behavior
The cluster should be still available, because only 2 out of 6 disks are down.

## Your Environment
* Version used (`minio version`):RELEASE.2017-11-22T19-55-46Z
* MC version: latest
* Environment: EC2

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
These deadlines are needed such that we fail pro-actively
for reach call beyond certain expected time limit. This
PR addresses an important problem when sudden server
shutdown could lead to perceived hangs of all the cluster
operations.

## Motivation and Context
Fixes #5290
Fixes #4887
Fixes #5092

## How Has This Been Tested?
Manually using reproducers provided in each of the bugs.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#4554a56f-f675-4585-99ef-599ee006554d) |
<a name="4554a56f-f675-4585-99ef-599ee006554d"></a>
### 5577-c64aa44/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30085
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ecf43233b99d:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-yevivdtymwbg10",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 348,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30085/aws-sdk-go-test-yevivdtymwbg10: dial tcp 192.168.122.45:30085: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#cc68d15f-8296-4501-abd7-d8a9d15e8b19) |
<a name="cc68d15f-8296-4501-abd7-d8a9d15e8b19"></a>
### 5577-8b87817/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30186
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 8b8d632bedf1:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-g3pqzbztj9t626",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 351,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30186/aws-sdk-go-test-g3pqzbztj9t626: dial tcp 192.168.122.45:30186: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#18a29207-7ee4-4394-b553-a550b5c49b43) |
<a name="18a29207-7ee4-4394-b553-a550b5c49b43"></a>
### 5577-1126964/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30635
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 3fd829fa5248:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ftz56mdl4ugwzi",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 4521,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30635/aws-sdk-go-test-ftz56mdl4ugwzi: dial tcp 192.168.39.45:30635: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

File: cmd/endpoint-ellipses.go [hunk 1]
```diff
@@ -165,9 +165,9 @@ func getTotalSizes(argPatterns []ellipses.ArgPattern) []uint64 {
 func parseEndpointSet(args ...string) (ep endpointSet, err error) {
 	var argPatterns = make([]ellipses.ArgPattern, len(args))
 	for i, arg := range args {
-		patterns, err := ellipses.FindEllipsesPatterns(arg)
-		if err != nil {
-			return endpointSet{}, err
+		patterns, perr := ellipses.FindEllipsesPatterns(arg)
+		if perr != nil {
+			return endpointSet{}, perr
```
Comment (balamurugana) [inline cmd/endpoint-ellipses.go]: Are these changes related to this PR?

File: cmd/auth-rpc-client.go [hunk 2]
```diff
@@ -127,11 +128,14 @@ func (authClient *AuthRPCClient) Login() (err error) {
...
 		var rpcClient *rpc.Client
-		rpcClient, err = rpcDial(authClient.config.serverAddr, authClient.config.serviceEndpoint, authClient.config.secureConn)
+		var conn net.Conn
+		rpcClient, conn, err = rpcDial(authClient.config.serverAddr, authClient.config.serviceEndpoint, authClient.config.secureConn)
 		if err != nil {
 			return err
 		}
 
+		// Set timeout for login.
+		conn.SetDeadline(time.Now().Add(defaultDialTimeout))
```
Comment (balamurugana) [inline cmd/auth-rpc-client.go]: Use `time.Now().UTC()` for consistency.

File: cmd/auth-rpc-client.go [hunk 3]
```diff
@@ -140,10 +144,13 @@ func (authClient *AuthRPCClient) Login() (err error) {
 			}
 			return err
 		}
+		// Flush timeout so that RPC readloop doesn't exit.
+		conn.SetDeadline(time.Time{})
```
Comment (balamurugana) [inline cmd/auth-rpc-client.go]: use `defer` to clear deadline

File: cmd/auth-rpc-client.go [hunk 4]
```diff
@@ -233,18 +247,17 @@ const connectSuccessMessage = "200 Connected to Go RPC"
 
 // dial tries to establish a connection to serverAddr in a safe manner.
 // If there is a valid rpc.Cliemt, it returns that else creates a new one.
-func rpcDial(serverAddr, serviceEndpoint string, secureConn bool) (netRPCClient *rpc.Client, err error) {
+func rpcDial(serverAddr, serviceEndpoint string, secureConn bool) (netRPCClient *rpc.Client, conn net.Conn, err error) {
```
Comment (balamurugana) [inline cmd/auth-rpc-client.go]: Returning `rpc.Client` and `net.Conn` look unclean.  You could return `net.Conn` and encapsulate  `rpc.Client` within `authClient`

File: cmd/storage-rpc-client.go [hunk 5]
```diff
@@ -311,5 +317,5 @@ func (n *networkStorage) RenameFile(srcVolume, srcPath, dstVolume, dstPath strin
 		SrcPath: srcPath,
 		DstVol:  dstVolume,
 		DstPath: dstPath,
-	}, &reply)
+	}, &reply, 10*time.Minute)
```
Comment (balamurugana) [inline cmd/storage-rpc-client.go]: 1. Have these timeouts been discussed including side effects of then majorly in write operations?  
2.  These timeouts need to be declared as constants with respective comments.

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: cmd/endpoint-ellipses.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/endpoint-ellipses.go]: Primarily because `go vet` started complaining suddenly. Have to do it outside of this PR.

File: cmd/storage-rpc-client.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/storage-rpc-client.go]: > Have these timeouts been discussed including side effects of then majorly in write operations?

Basing these timeouts based on existing timeouts for Dsync operations, will move them as constants.

File: cmd/auth-rpc-client.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/auth-rpc-client.go]: Sure will do that.

**Comment by harshavardhana** [comment]
PR updated @balamurugana @krishnasrinivas - regarding the timeout values its blocked for now on @abperiasamy

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

File: cmd/storage-rpc-client.go [hunk 6]
```diff
@@ -32,6 +33,15 @@ type networkStorage struct {
 	connected bool
 }
 
+// List of Call timeouts for each class of storage operations
+const (
+	volumeOpTimeout        = 10 * time.Second
+	volumeListOpTimeout    = 10 * time.Minute
+	fileStatOpTimeout      = 10 * time.Second
+	fileReadWriteOpTimeout = 10 * time.Minute
+	fileListOpTimeout      = 30 * time.Minute
```
Comment (vadmeste) [inline cmd/storage-rpc-client.go:42]: @harshavardhana, I think having large timeout values doesn't fix the problem of clients wondering why it takes too much time to list a bucket contents when one node fails. I think to fix this use case, we need to fork net/rpc and we see why even when KeepAlive timeout is set, rpc.Call() doesn't quit when one node fails.

Comment (harshavardhana) [reply cmd/storage-rpc-client.go:42]: The issue is actually not in rpc.Call() @vadmeste the issue is basically in net.Conn which was previously remembered doesn't return or timeout. Without deadlines Go has no timeouts for net.Conn

**Comment by balamurugana** [comment]
```
Error running mint automation
+ mkdir -p 5577-3116ba0/gopath/src/github.com/minio
+ git -C 5577-3116ba0/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 5577-3116ba0/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 5577-3116ba0/gopath/src/github.com/minio/minio fetch --quiet minio pull/5577/head:pr5577
+ git -C 5577-3116ba0/gopath/src/github.com/minio/minio checkout --quiet pr5577
+ GOPATH=/home/bala/mintauto/5577-3116ba0/gopath
+ make -C 5577-3116ba0/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/auth-rpc-client.go:312:15: undefined: "github.com/minio/minio/pkg/http".ReadResponse
cmd/auth-rpc-client.go:312:57: undefined: "github.com/minio/minio/pkg/http".Request
make: *** [build] Error 2

```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#251b95b3-3e0c-4ccf-a73b-1513429eec63) |
<a name="251b95b3-3e0c-4ccf-a73b-1513429eec63"></a>
### 5577-299ad7a/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30293
ACCESS_KEY:      minioazure1
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 4259fbc2a617:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... done in 6 minutes and 55 seconds
(3/12) Running aws-sdk-ruby tests ... done in 1 minutes and 3 seconds
(4/12) Running awscli tests ... FAILED in 5 minutes and 22 seconds
{
  "name": "awscli",
  "duration": 6110,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-17146/datafile-1-MBremove_bucket: awscli-mint-test-bucket-17146"
}

Executed 3 out of 12 tests successfully.
```

**Comment by vadmeste** [comment]
@harshavardhana tested this PR with a node forcefully shutdown.  In practice, listing will be stuck for 10 minutes but then everything will come back to normal. The tcp keepalive timeout (10 seconds) is not working. I think we can address that in another PR.

**Comment by harshavardhana** [comment]
> @harshavardhana tested this PR with a node forcefully shutdown. In practice, listing will be stuck for 10 minutes but then everything will come back to normal. The tcp keepalive timeout (10 seconds) is not working. I think we can address that in another PR.

The major caveat in current approach is that each RPC is required to finish within the required timeout value, but it is possible that this value might not be right. But trying to add deadlines exactly like our `pkg/http` package doesn't work since `net/rpc` runs a Read() loops and our net.Conn will always timeout if we intend to keep a lower level value. 

Need to update @abperiasamy again that this approach doesn't work and the approach taken in this PR is the only approach right now otherwise we will have to re-implement `net/rpc`.

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#dc4cee94-f6b8-4370-8d4d-c6f1e461e7d9) |
<a name="dc4cee94-f6b8-4370-8d4d-c6f1e461e7d9"></a>
### 5577-1055151/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32636
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ba5d562acc4c:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 4 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-k3bf5l6ywbgzng",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 4503,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:32636/aws-sdk-go-test-k3bf5l6ywbgzng: dial tcp 192.168.122.45:32636: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#55724e40-c721-4923-abbf-36411facc995) |
<a name="55724e40-c721-4923-abbf-36411facc995"></a>
### 5577-f48a1ef/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32187
ACCESS_KEY:      minioazure1
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp cab439ca301b:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... done in 8 minutes and 13 seconds
(3/12) Running aws-sdk-ruby tests ... done in 50 seconds
(4/12) Running awscli tests ... done in 3 minutes and 39 seconds
(5/12) Running mc tests ... done in 8 minutes and 15 seconds
(6/12) Running minio-dotnet tests ... done in 1 minutes and 18 seconds
(7/12) Running minio-go tests ... done in 1 seconds
(8/12) Running minio-java tests ... done in 32 seconds
(9/12) Running minio-js tests ... FAILED in 2 minutes and 37 seconds
{
  "name": "minio-js",
  "function": "getBucketPolicy(bucketName, cb)",
  "args": "bucketName:minio-js-test-c93cabd0-f7a1-4c66-ac58-0a3c9e008c8a",
  "duration": 263,
  "status": "FAIL",
  "error": "S3Error: The bucket policy does not exist at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:56:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:117:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:48:26) at DestroyableTransform.<anonymous> (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:135:12) at finishMaybe (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:371:12) at endWritable (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:378:3) at DestroyableTransform.Writable.end (node_modules/minio/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:356:5) at IncomingMessage.onend (_stream_readable.js:511:10) at endReadableNT (_stream_readable.js:974:12) at _combinedTickCallback (internal/process/next_tick.js:80:11) at process._tickCallback (internal/process/next_tick.js:104:9)"
}

Executed 8 out of 12 tests successfully.
```

**Comment by harshavardhana** [comment]
Closing this PR since we will be implementing this differently by re-implementing our net/rpc requirements

**Comment by starinacool** [comment]
Hi

Have this issue was fixed the other way?
I faced this problem when I switched one of my 4 nodes off.

Had to do this tuning on nodes:

net.ipv4.tcp_syn_retries = 1
net.ipv4.tcp_retries1 = 1
net.ipv4.tcp_retries2 = 1

After this delay droped to 3 sec when i try to put an object.

Constant pinging via icmp and disabling a node for some time would solve this problem I think.
